### ISSUE TITLE
Identity Management codeunit in extensions

### ISSUE THREAD

**Issue by kvaldik** [Microsoft/AL#911]
Hi all.

As discussed in [#496](https://github.com/Microsoft/AL/issues/496) the identity management codeunit is not available to extensions, but is it completely off the table?

As I describe in the linked issue we have built a payment app that we invoke from our D365 extension using URI. Included in the URI is a link to a D365 web service, the username and password (web service access key). The payment app collects the payment and posts the results to D365 through the provided web service.

It would be a lot nicer to be able to do this using the web service access key instead of having to implement the OAuth functionality in our payment app. The user experience is also better when there is no need enter the credentials in our app.

**Comment by kvaldik** [comment]
The functions that we would need are:

GetWebServiceExpiryDate
CreateWebServicesKey
GetWebServicesKey

**Comment by SBalslev** [comment]
Currently considering your feedback. While it seems innocent to open up for existing functionality it does have consequences and the more connected the service gets the harder it will become to support legacy authentication using webservicekey. Without an AAD token from the user then even a simple scenario like updating a contact name could result in other features and functionality not working, unpredictable behavior or even worse inconsistent data.

**Comment by SBalslev** [comment]
After careful consideration and discussion with our security experts, we have decided not to expose the web serviceskey functions to extensions.
We have created a new feature description extending the current functionality to be web services key pr. extension. This will allow us to expose the web services key like functionality to extensions. The feature is currently planned to finish in the spring of 2018.
If you have ideas or comments for such a feature feel free to send them to me.

For now, the best way to talk back into financials remains using AAD tokens acquired and maintained by redirecting the user to your custom web application.

References:
See "Web Application to Web API" section here: https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-authentication-scenarios
Web API sample: https://github.com/Azure-Samples/active-directory-dotnet-webapi-onbehalfof
Web App to Web API sample: https://github.com/Azure-Samples/active-directory-dotnet-webapp-webapi-multitenant-openidconnect
Token Cache sample: https://docs.microsoft.com/en-us/azure/architecture/multitenant-identity/token-cache
